keps/sig-network/2732-np-version/README.md

```diff
@@ -0,0 +1,803 @@
+features the NetworkPolicy implementation supports". Especially, there
+is often a lag between when new NetworkPolicy features are defined and
+when they are widely implemented. This means that in some cases a user
+cannot know if a cluster will correctly process a given NetworkPolicy
+(and makes it difficult to add new NetworkPolicy features if we want
+to ensure they won't be catastrophically misinterpreted by older
+implementations).
+
+```
+<<[UNRESOLVED snowflake ]>>
+
+Just how special/unusual is NetworkPolicy in this regard? Have any
+other SIGs already dealt with similar problems? Or if not are there
+any other APIs that are similarly failing to deal with the same
+problems?
+
+<<[/UNRESOLVED]>>
+```
+
+Somewhat related to this, there is currently no way for a
+NetworkPolicy implementation (ie, the network plugin) to indicate the
+status of NetworkPolicy processing to the user who created the policy.
+
+This KEP attempts to solve both of these problems.
+
+(Although it is not technically required that NetworkPolicy be
+implemented by the network plugin, this is almost universally the case,
+and so I will refer to NetworkPolicy implementations as "network
+plugins" through the rest of this document.)
+
+## Motivation
+
+### Goals
+
+- Provide a better mechanism for dealing with version skew with
+  respect to NetworkPolicy features. Specifically:
+
+    - Allow the apiserver to indicate to the network plugin what
+      version of the NetworkPolicy featureset a particular policy
+      requires (eg, whether it uses combined `podSelector` and
+      `namespaceSelector`), so that network plugins can reliably
+      recognize when a NetworkPolicy makes use of a feature that the
+      NetworkPolicy implementation does not support.
+
+    - Allow NetworkPolicy implementations to indicate back to the user
+      when they are aware that they cannot correctly implement a given
+      NetworkPolicy.
+
+- Define rules for dealing with feature gates and alpha APIs in
+  NetworkPolicy that work well with the 3-way versioning split.
+
+- Allow network plugins to indicate when a NetworkPolicy has been fully
+  "programmed" into the cluster network, to allow clients (including
+  humans, scripts, and cluster components) to get more reliable behavior
+  when adding new policies.
+
+- (Maybe) Allow network plugins to provide some sort of global status
+  object with certain information about the implementation.
+
+### Non-Goals
+
+- Deprecating (or requiring) the use of feature gates to indicate the
+  stability / "certainty" of NetworkPolicy API additions; whether a
+  feature is API-stable or not is distinct from whether it is widely
+  implemented by plugins or not (though there is generally bidirectional
+  influence between the two).
+
+- Defining "metric-like" NetworkPolicy status information (eg, how long
+  a particular rule took to implement).
+
+## Proposal
+
+### User Stories
+
```

Comment:
I would add a first “story 0”:
as a developer, I want to find out if NetworkPolicy applies _at all_ in my cluster. It's plausible that the API is present but the cluster network pays it no heed at all.

Reply:
@sftim That is a good question. The objective of this KEP should unify the community around network policy in a future version since there are differences in the features and scope of network policy across various CNIs. The implied issue is whether there would be an opt so "not asserted" would be an option in the API.

Reply:
> I would add a first “story 0”:
> as a developer, I want to find out if NetworkPolicy applies at all in my cluster.

That's sort of a special case of Story 3; for when the plugin doesn't implement _any_ NetworkPolicy features